bolt-modules/prompt/lib/puppet/functions/prompt/menu.rb

```diff
@@ -0,0 +1,103 @@
+# frozen_string_literal: true
+
+require 'bolt/error'
+
+# Display a menu prompt and wait for a response. Continues to prompt
+# until an option from the menu is selected.
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:'prompt::menu') do
+  # Select from a list of options.
+  # @param prompt The prompt to display.
+  # @param menu A list of options to choose from.
+  # @param options A hash of additional options.
+  # @option options [String] default The default option to return if the user does not provide
+  #   input or if standard in (stdin) is not a tty. Must be an option present in the menu.
+  # @return The selected option.
+  # @example Prompt the user to select from a list of options
+  #   $selection = prompt::menu('Select a fruit', ['apple', 'banana', 'carrot'])
+  # @example Prompt the user to select from a list of options with a default value
+  #   $selection = prompt::menu('Select environment', ['development', 'production'], 'default' => 'development')
+  dispatch :prompt_menu_array do
+    param 'String', :prompt
+    param 'Array[Any]', :menu
+    optional_param 'Hash[String[1], Any]', :options
```

Comment:
Should this be `String[1], String`?

Reply:
That would limit the values that you can choose from to strings, which I don't see a good reason for doing. Only the keys need to be strings, since you need to be able to type the input.

Reply:
Hm, I guess since the menu can accept any types we're not beholden to what the user types in (like with `prompt`), so this actually could be `Any`... :thinking: I think it'd be simpler if everything accepted strings, but I don't think that's a good enough reason to limit it since I'm not sure what we'd do different once someone asked for this feature.

Reply:
Heh, whoops, ignore that last comment. I was thinking this was for the `menu` param. Since the values can be anything, the `default` option also needs to have the `Any` type.

Reply:
Hmm, I actually think limiting this to `Data` instead of `Any` makes perfect sense - I don't understand the use case where people would want any other types, and that prevents this from having too much undefined behavior.

Reply:
See https://github.com/puppetlabs/bolt/pull/2714#discussion_r595563927